docs/contributors/code/coding-guidelines.md

```diff
@@ -136,10 +136,278 @@ export { __unstableDoTerribleAwfulAction } from './api';
+	// MyComponent.js:
+	function hasActiveBlockOverlayActive ( selectors, parent ) { /* ... */ }
+	function useIsWithinBlockOverlay( clientId ) {
+		return useSelect( ( select ) => {
+			const selectors = select( blockEditorStore ); 
+			let parent = selectors.getBlockRootClientId( clientId );
+			while ( !!parent ) {
+				if ( hasActiveBlockOverlayActive( selectors, parent ) ) {
+					return true;
+				}
+				parent = selectors.getBlockRootClientId( parent );
+			}
+			return false;
+		});
+	}
+	function MyComponent({ clientId }) {
+		const isWithinBlockOverlay = useIsWithinBlockOverlay( clientId );
+		// ...
+	}
+```
+
+If your experimental selector needs to access a part of the store that isn't
+accessible by the existing public selectors, you may access the store data directly
+via `useRegistry()`:
+
+```js
+function MyComponent() {
+	const registry = useRegistry();
+	const enabledFeatures = registry.stores['my-store'].store.getState().features;
+	// ...
+}
+```
+
+#### Replace experimental actions with hooks and thunks
+
+Public experimental actions can be replaced with non-exported React hooks:
+
+```js
+// Instead of this:
+	// actions.js:
+	export function __experimentalBeforeSave() {
+		return { type: 'BEFORE_SAVE' };
+	}
+	// MyComponent.js:
+	function MyComponent() {
+		const { __experimentalBeforeSave } = useDispatch( myStore );
+		const onClick = () => __experimentalBeforeSave();
+		// ...
+	}
+
+// Consider this:
+	// MyComponent.js:
+	function MyComponent() {
+		const registry = useRegistry();
+		const onClick = () => registry.stores['my-store'].store.dispatch({ type: 'BEFORE_SAVE' });
+		// ...
+	}
+```
+
+In addition, the existing public actions can dispatch private actions:
+
+```js
+export function toggleFeature( scope, featureName ) {
+    return function ( { dispatch } ) {
+		dispatch({ type: 'BEFORE_SAVE' })
+		// ...
+    };
+}
+```
+
+#### `lock()` and `unlock()`
+
+Use the `lock()` and `unlock()` API from `@wordpress/experiments` to
+privately export almost anything. See the examples below and learn more
+in the [`@wordpress/experiments` package README.md](https://github.com/WordPress/gutenberg/blob/HEAD/packages/experiments/README.md).
+
+##### Experimental selectors and actions
+
+Attach private selectors and actions to a public store:
+
+```js
+// In @wordpress/package1/store.js:
+import { __experimentalHasContentRoleAttribute } from './selectors';
+import { lock } from './experiments';
+
+export const store = registerStore(/* ... */);
+/* Attach a private selector to the exported store: */
+lock(store, {
+	__experimentalHasContentRoleAttribute
+});
+
+// In @wordpress/package2/MyComponent.js:
+import { store } from '@wordpress/package1';
+import { useSelect } from '@wordpress/data';
+import { unlock } from './experiments';
+
+function MyComponent() {
+    const hasRole = useSelect( ( select ) => (
+		// Use the private selector:
+        unlock( select( store ) ).__experimentalHasContentRoleAttribute()
```

Comment:
My understanding is that `unlock` can only be called once for a given package, this is going to be called several times, is it going to work still? If yes, I don't see how this stops any third party code from accessing these selectors/actions.

Reply:
You can call `unlock()` as many times as you want. The catch is that the only way to access the `unlock` function is by calling the `__dangerousOptInToUnstableAPIsOnlyForCoreModules` function:

```js
// packages/block-editor/experiments.js
import { __dangerousOptInToUnstableAPIsOnlyForCoreModules } from '@wordpress/experiments';
const { lock, unlock } = __dangerousOptInToUnstableAPIsOnlyForCoreModules(
	'<CONSENT STRING>',
	'@wordpress/blocks'
);

// packages/block-editor/my-component.js
import { unlock } from './experiments';

unlock( select( store ) ). __experimentalHasContentRoleAttribute()
unlock( select( store ) ). __experimentalHasContentRoleAttribute()
```

I'll describe it more explicitly

Reply:
Ok makes sense, I guess that means "selectors and actions" are considered solved by this API 👍 (unlike props, arguments, theme.json, block.json)